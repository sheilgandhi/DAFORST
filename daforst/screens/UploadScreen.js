import {
  View,
  Text,
  StyleSheet,
  SafeAreaView,
  TextInput,
  ScrollView,
  useWindowDimensions,
  Pressable,
  NativeModules,
} from 'react-native';
import React, { useState } from 'react';
import IPFS from 'ipfs-mini';
import { COLORS } from '../colors';
import axios from 'axios';

const UploadScreen = ({ navigation }) => {
  const [image, setImage] = useState(null);

  const [id, setId] = useState('');
  const [location, setLocation] = useState('');
  const [ipfsCid, setIpfsCid] = useState('');

  const [name, setName] = useState('');
  const [owner, setOwner] = useState('');
  const [desc, setDesc] = useState('');

  const { width } = useWindowDimensions();

  const ipfs = new IPFS({
    host: 'ipfs.infura.io',
    port: 5001,
    protocol: 'https',
  });

  /**
   * Uses React Native Bridge
   */
  const handleArCore = () => {
    NativeModules.ARCore.startARCore();
  };

  const uploadToIpfs = () => {
    ipfs
      .add(image.assets[0].base64)
      .then(hash => {
        console.log(hash);
        setIpfsCid(hash);
      })
      .catch(console.log);
  };

  const postToDb = async () => {
    await axios
      .post('http://10.0.2.2:3000/api/assets/', {
        arcore_id: id,
        arcore_location: location,
        ipfs_cid: ipfsCid,
        name: name,
        owner: owner,
        description: desc,
      })
      .then(function (response) {
        console.log(response);
      })
      .catch(function (error) {
        console.log(error);
      });
  };

  const handleAddAsset = () => {
    console.log(image.assets[0].base64);

    // ARCore here

    // IPFS
    uploadToIpfs();
    // Db Persistence
    postToDb();
    // navigate user to assets
    navigation.navigate('Assets');
  };

  const handleClear = () => {
    setImage(null);
    setId('');
    setLocation('');
    setIpfsCid('');
    setName('');
    setOwner('');
    setDesc('');
  };

  return (
    <SafeAreaView style={styles.container}>
      <ScrollView>
        {/* Upload */}
        <View style={styles.uploadGroup}>
          <Pressable onPress={handleArCore} style={styles.upload}>
            <Text style={{ color: COLORS.black }}>ðŸ“· Capture</Text>
          </Pressable>
        </View>
        {/* ARCore */}
        <View>
          <Text testID="id" style={styles.label}>
            ID: {id || '...'}
          </Text>
          <Text testID="location" style={styles.label}>
            Location: {location || '...'}
          </Text>
        </View>
        {/* Form Basic */}
        <View>
          <View style={styles.inputGroup}>
            <Text style={styles.label}>Name</Text>
            <TextInput
              testID="name"
              placeholder="Name"
              style={styles.input}
              inputStyle={{ color: COLORS.black }}
              placeholderTextColor={COLORS.black}
              onChangeText={setName}
              value={name}
            />
          </View>
          <View style={styles.inputGroup}>
            <Text style={styles.label}>Owner</Text>
            <TextInput
              testID="owner"
              placeholder="Owner"
              style={styles.input}
              inputStyle={{ color: COLORS.black }}
              placeholderTextColor={COLORS.black}
              onChangeText={setOwner}
              value={owner}
            />
          </View>
          <View style={styles.inputGroup}>
            <Text style={styles.label}>Description</Text>
            <TextInput
              testID="description"
              placeholder="Description of asset"
              style={[
                styles.input,
                { minHeight: 75, textAlignVertical: 'top' },
              ]}
              inputStyle={{ color: COLORS.black }}
              placeholderTextColor={COLORS.black}
              onChangeText={setDesc}
              value={desc}
              multiline
            />
          </View>
        </View>
        {/* Form Advanced */}
        <View></View>
        {/* Buttons */}
        <View style={styles.buttons}>
          <Pressable
            testID="uploadButton"
            onPress={handleAddAsset}
            style={[
              styles.button,
              { backgroundColor: COLORS.green, width: width - 64 },
            ]}>
            <Text style={[styles.buttonText, { color: COLORS.white }]}>
              Upload
            </Text>
          </Pressable>
          <Pressable
            testID="clearButton"
            onPress={handleClear}
            style={[
              styles.button,
              { backgroundColor: COLORS.white, width: width - 64 },
            ]}>
            <Text style={[styles.buttonText, { color: COLORS.black }]}>
              Clear
            </Text>
          </Pressable>
        </View>
      </ScrollView>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    marginHorizontal: 24,
  },
  uploadGroup: {
    marginVertical: 20,
  },
  upload: {
    padding: 10,
    backgroundColor: COLORS.grey,
    borderRadius: 10,
    justifyContent: 'center',
    alignItems: 'center',
    marginBottom: 10,
  },
  inputGroup: {
    marginTop: 15,
  },
  label: {
    color: COLORS.green,
    paddingBottom: 10,
    fontWeight: 'bold',
    fontSize: 16,
  },
  input: {
    backgroundColor: COLORS.grey,
    color: COLORS.black,
    borderRadius: 10,
    paddingLeft: 20,
  },
  buttons: {
    justifyContent: 'space-evenly',
    alignItems: 'center',
    paddingBottom: 10,
  },
  button: {
    height: 50,
    borderRadius: 96,
    justifyContent: 'center',
    alignItems: 'center',
    margin: 10,
    borderColor: COLORS.green,
    borderWidth: 3,
  },
  buttonText: {
    fontSize: 18,
    textAlign: 'center',
  },
});

export default UploadScreen;
