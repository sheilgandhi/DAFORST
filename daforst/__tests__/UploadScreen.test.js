import React from 'react';
import renderer from 'react-test-renderer';
import UploadScreen from '../screens/UploadScreen';
import {
  fireEvent,
  render,
  waitFor,
  cleanup,
} from '@testing-library/react-native';
import { faker } from '@faker-js/faker';

const navigation = {
  goBack: jest.fn(),
  navigate: jest.fn(),
  setOptions: jest.fn(),
};

describe('UploadScreen', () => {
  afterEach(() => {
    cleanup();
  });

  test('renders correctly', () => {
    const tree = renderer
      .create(<UploadScreen navigation={navigation} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('validate name input fills', async () => {
    const name = faker.name.findName();
    const { getByTestId } = render(<UploadScreen navigation={navigation} />);
    const input = getByTestId('name');
    expect(input).toBeTruthy(); // exists?
    fireEvent.changeText(input, name);
    expect(input.props.value).toBe(name);
  });

  test('validate owner input fills', async () => {
    const owner = faker.name.findName();
    const { getByTestId } = render(<UploadScreen navigation={navigation} />);
    const input = getByTestId('owner');
    expect(input).toBeTruthy(); // exists?
    fireEvent.changeText(input, owner);
    expect(input.props.value).toBe(owner);
  });

  test('validate description multiline input fills', async () => {
    const description = faker.lorem.lines();
    const { getByTestId } = render(<UploadScreen navigation={navigation} />);
    const input = getByTestId('description');
    expect(input).toBeTruthy(); // exists?
    fireEvent.changeText(input, description);
    expect(input.props.value).toBe(description);
  });
});
