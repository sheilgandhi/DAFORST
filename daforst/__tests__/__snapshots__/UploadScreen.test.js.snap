// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UploadScreen renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
      "marginHorizontal": 24,
    }
  }
>
  <RCTScrollView>
    <View>
      <View
        style={
          Object {
            "marginVertical": 20,
          }
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#f9f9f9",
              "borderRadius": 10,
              "justifyContent": "center",
              "marginBottom": 10,
              "padding": 10,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#000",
              }
            }
          >
            üñºÔ∏è Photo Gallery
          </Text>
        </View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#f9f9f9",
              "borderRadius": 10,
              "justifyContent": "center",
              "marginBottom": 10,
              "padding": 10,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#000",
              }
            }
          >
            üì∑ Camera
          </Text>
        </View>
      </View>
      <View>
        <Text
          style={
            Object {
              "color": "#55897B",
              "fontSize": 16,
              "fontWeight": "bold",
              "paddingBottom": 10,
            }
          }
          testID="id"
        >
          ID: 
          ...
        </Text>
        <Text
          style={
            Object {
              "color": "#55897B",
              "fontSize": 16,
              "fontWeight": "bold",
              "paddingBottom": 10,
            }
          }
          testID="location"
        >
          Location: 
          ...
        </Text>
      </View>
      <View>
        <View
          style={
            Object {
              "marginTop": 15,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#55897B",
                "fontSize": 16,
                "fontWeight": "bold",
                "paddingBottom": 10,
              }
            }
          >
            Name
          </Text>
          <TextInput
            inputStyle={
              Object {
                "color": "#000",
              }
            }
            onChangeText={[Function]}
            placeholder="Name"
            placeholderTextColor="#000"
            style={
              Object {
                "backgroundColor": "#f9f9f9",
                "borderRadius": 10,
                "color": "#000",
                "paddingLeft": 20,
              }
            }
            testID="name"
            value=""
          />
        </View>
        <View
          style={
            Object {
              "marginTop": 15,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#55897B",
                "fontSize": 16,
                "fontWeight": "bold",
                "paddingBottom": 10,
              }
            }
          >
            Owner
          </Text>
          <TextInput
            inputStyle={
              Object {
                "color": "#000",
              }
            }
            onChangeText={[Function]}
            placeholder="Owner"
            placeholderTextColor="#000"
            style={
              Object {
                "backgroundColor": "#f9f9f9",
                "borderRadius": 10,
                "color": "#000",
                "paddingLeft": 20,
              }
            }
            testID="owner"
            value=""
          />
        </View>
        <View
          style={
            Object {
              "marginTop": 15,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#55897B",
                "fontSize": 16,
                "fontWeight": "bold",
                "paddingBottom": 10,
              }
            }
          >
            Description
          </Text>
          <TextInput
            inputStyle={
              Object {
                "color": "#000",
              }
            }
            multiline={true}
            onChangeText={[Function]}
            placeholder="Description of asset"
            placeholderTextColor="#000"
            style={
              Array [
                Object {
                  "backgroundColor": "#f9f9f9",
                  "borderRadius": 10,
                  "color": "#000",
                  "paddingLeft": 20,
                },
                Object {
                  "minHeight": 75,
                  "textAlignVertical": "top",
                },
              ]
            }
            testID="description"
            value=""
          />
        </View>
      </View>
      <View />
      <View
        style={
          Object {
            "alignItems": "center",
            "justifyContent": "space-evenly",
            "paddingBottom": 10,
          }
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "borderColor": "#55897B",
                "borderRadius": 96,
                "borderWidth": 3,
                "height": 50,
                "justifyContent": "center",
                "margin": 10,
              },
              Object {
                "backgroundColor": "#55897B",
                "width": 686,
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontSize": 18,
                  "textAlign": "center",
                },
                Object {
                  "color": "#fff",
                },
              ]
            }
          >
            Upload
          </Text>
        </View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "borderColor": "#55897B",
                "borderRadius": 96,
                "borderWidth": 3,
                "height": 50,
                "justifyContent": "center",
                "margin": 10,
              },
              Object {
                "backgroundColor": "#fff",
                "width": 686,
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontSize": 18,
                  "textAlign": "center",
                },
                Object {
                  "color": "#000",
                },
              ]
            }
          >
            Clear
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
